main/lsp/TimeTravelingGlobalState.h

```diff
@@ -0,0 +1,111 @@
+#ifndef RUBY_TYPER_LSP_TIMETRAVELINGGLOBALSTATE_H
+#define RUBY_TYPER_LSP_TIMETRAVELINGGLOBALSTATE_H
+
+#include "common/concurrency/WorkerPool.h"
+#include "common/kvstore/KeyValueStore.h"
+#include "core/NameHash.h"
+#include "core/core.h"
+#include "main/lsp/LSPConfiguration.h"
+#include "main/lsp/json_types.h"
+#include <deque>
+
+namespace sorbet::realmain::lsp {
+
+/**
+ * A class that encloses a global state and allows one to 'time-travel' it between versions.
+ * It also time-travels an array of file hashes corresponding to the files in the time-traveled global state.
+ * Updates cannot be re-ordered, so updates that are undone are assumed to eventually become re-applied.
+ */
+class TimeTravelingGlobalState final {
+private:
+    // Contains file updates and their respective hash updates.
+    struct GlobalStateUpdate {
+        std::vector<std::shared_ptr<core::File>> fileUpdates;
+        std::vector<core::FileHash> hashUpdates;
+    };
+
+    /**
+     * An undo/redo log entry for message with internal id `messageId`.
+     * Applying the updates to initialGS will apply or undo the edit with that messageId, depending on if it is in the
+     * undo or redo log.
+     */
+    struct TimeTravelUpdate {
+        int version = 0;
+        GlobalStateUpdate update;
+        GlobalStateUpdate undoUpdate;
+    };
+    const LSPConfiguration config;
+    std::shared_ptr<spdlog::logger> logger;
+    WorkerPool &workers;
+    std::unique_ptr<KeyValueStore> kvstore; // always null for now.
+    std::unique_ptr<core::GlobalState> gs;
+
+    // Indicates the current version of `gs`.
```

Comment:
reading comments for these two version, the distinction between the two wasn't clear for me.

Reply:
One is the current version of `gs`.

The other is the latest file update we've seen.

I can update the text to be clearer.